### PR TITLE
Implement Atomic Blob Writes for HDFS Repository

### PR DESCRIPTION
**Pull request by original-brownbear**
* Implement atomic writes the same way we do for the `FsBlobContainer` via rename which is atomic
* Relates #37011

### CODE REVIEW

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java [hunk 1]
```diff
@@ -116,6 +118,30 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize, b
...
+        store.execute((Operation<Void>) fileContext -> {
+            final Path tempBlobPath = new Path(path, tempBlob);
+            try (FSDataOutputStream stream =
+                     fileContext.create(tempBlobPath, EnumSet.of(CreateFlag.CREATE),  CreateOpts.bufferSize(bufferSize))) {
+                int bytesRead;
+                byte[] buffer = new byte[bufferSize];
+                while ((bytesRead = inputStream.read(buffer)) != -1) {
+                    stream.write(buffer, 0, bytesRead);
+                }
+                stream.hsync();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java]: Created https://github.com/elastic/elasticsearch/pull/37069 to speed up writes using the same strategy for non-atomic writes.

**Comment by original-brownbear** [comment]
Fails tests for some reason, looking into it

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (original-brownbear) [reply plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java]: Urgh, you're right ... should've read the second paragraph in the docs of that method :)
Your solution is better will adjust.

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java [hunk 2]
```diff
@@ -116,6 +118,30 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize, b
         });
     }
 
+    @Override
+    public void writeBlobAtomic(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+        final String tempBlob = FsBlobContainer.tempBlobName(blobName);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java]: I'd rather not I think. This method is kinda specific to the two file-system based implementations. Not sure it makes sense blowing up the interface for that?

**Comment by original-brownbear** [comment]
@ywelsch fixed the sync method :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Jenkins runs gradle build tests 1 please

**Comment by original-brownbear** [comment]
Jenkins run gradle build tests 1

**Pull request merged**